Userland/Libraries/LibGfx/ImageFormats/AVIFLoader.cpp

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (c) 2023, Nico Weber <thakis@chromium.org>
+ * Copyright (c) 2024, doctortheemh <doctortheemh@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Error.h>
+#include <LibGfx/ImageFormats/AVIFLoader.h>
+
+#include <avif/avif.h>
+
+namespace Gfx {
+
+struct AVIFLoadingContext {
+    enum State {
+        NotDecoded = 0,
+        Error,
+        HeaderDecoded,
+        BitmapDecoded,
+    };
+
+    State state { State::NotDecoded };
+    ReadonlyBytes data;
+    Optional<IntSize> size;
+    bool has_alpha { false };
+    size_t image_count { 0 };
+    size_t repetition_count { 0 };
+
+    Vector<ImageFrameDescriptor> frame_descriptors;
+};
+
+AVIFImageDecoderPlugin::AVIFImageDecoderPlugin(ReadonlyBytes data, OwnPtr<AVIFLoadingContext> context)
+    : m_context(move(context))
+{
+    m_context->data = data;
+}
+
+AVIFImageDecoderPlugin::~AVIFImageDecoderPlugin() = default;
+
+static ErrorOr<void> decode_avif_header(AVIFLoadingContext& context)
+{
+    if (context.state >= AVIFLoadingContext::HeaderDecoded)
+        return {};
+
+    avifDecoder* decoder = avifDecoderCreate();
+    if (decoder == NULL) {
+        return Error::from_string_literal("failed to allocate AVIF decoder");
+    }
+
+    avifResult result = avifDecoderSetIOMemory(decoder, context.data.data(), context.data.size());
+    if (result != AVIF_RESULT_OK) {
+        avifDecoderDestroy(decoder);
+        return Error::from_string_literal("Cannot set IO on avifDecoder");
+    }
+
+    result = avifDecoderParse(decoder);
+    if (result != AVIF_RESULT_OK) {
+        avifDecoderDestroy(decoder);
+        return Error::from_string_literal("Failed to decode AVIF");
+    }
+
+    // image header decoded, save some results in context
+    context.size = IntSize { decoder->image->width, decoder->image->height };
+    context.has_alpha = decoder->alphaPresent == 1;
+    context.image_count = decoder->imageCount;
+    // FIXME: this could be negative
+    context.repetition_count = decoder->repetitionCount;
```

Comment:
I suppose for negative numbers you can set `repetition_count = 0` which is LibGfx's encoding of infinite. (Of course that somewhat ignores `AVIF_REPETITION_COUNT_UNKNOWN`, but the header says that's up to the application, so infinite would be a valid interpretation)